gist/gist

```diff
@@ -0,0 +1,812 @@
+  new_record=$(http_method PATCH "$GITHUB_API/gists/$GIST_ID" \
+               | sed -e '1 s/^/[/; $ s/$/]/' \
+               | _parse_response "${1#[[:alpha:]]}" )
+  [[ -n $new_record ]] && sed -i'' -E -e "/^$1 / s^.+^$new_record^" $INDEX \
+  && hint=false mark="$1 " _show_list \
+  || echo 'Fail to modify gist description'
+}
+
+# Print helper message
+usage() {
+  sed -E -n -e ' /^$/ q; 7,$ s/^# //p' "$0"
+}
+
+# Check remote urls of all repos match current protocol in configuration file
+# If not, update them
+_check_protocol() {
+  find $folder -maxdepth 1 -mindepth 1 -type d \
+  | while read -r repo; do
+    cd "$repo" || exit 1
+    url=$(git remote get-url origin)
+    if [[ $protocol == 'ssh' && $url =~ ^https ]]; then
+      git remote set-url origin "git@gist.github.com:$(basename $(pwd)).git"
+    elif [[ $protocol == 'https' && $url =~ ^git ]]; then
+      git remote set-url origin "https://gist.github.com/$(basename $(pwd)).git"
+    fi
+  done
+}
+
+_tag_gist() {
+  # if user want to change tags of a gist
+  if [[ -z $@ ]]; then
+    tag=true _show_list
+  elif _gist_id $1 &>/dev/null; then
+    _show_detail $1 | sed 3,6d && echo
+    local desc="$(_get_desc $1)"
+    local hashtags=$(_hashtags "$desc")
+    read -e -p 'Edit tags: ' -i "${hashtags//'#'/}" -r -a new_tags < /dev/tty
+    local new_hashtags="${new_tags[@]/#/#}"
+    local new_desc=$(sed "s/$hashtags$//; s/ *$/ /" <<<"$desc")${new_hashtags}
+    (_edit_gist $1 "$new_desc" &>/dev/null &)
+  # if user want to filter gists with given tags
+  else
+    local pattern="($(sed -E 's/([^ ]+)/#\1/g; s/ /[[:space:]]|/g' <<<"$@") )"
+    hint=false tag=true _show_list | grep --color=always -E "$pattern"
+  fi
+}
+
+# show all tags and pinned tags
+_show_tags() {
+  local pinned_tags=( $pin )
+  local tags=$(while read -r "${INDEX_FORMAT[@]}"; do
+               _hashtags "$description" | tr ' ' '\n' | sed -e '/^$/d'
+              done < $INDEX | sort -u) 
+
+  for prefix in {0..9} {a..z} {A-Z} [^0-9a-zA-Z]; do
+    local line=$(echo $tags | grep -Eo "#$prefix[^ ]+" | tr '\n' ' ')
+    [[ -z $line ]] && continue
+
+    # add color to pinned tags
+    echo -e "$(_color_pinned_tags "$line")"
+  done
+
+  echo
+  if [[ ${#pinned_tags} == 0 ]]; then
+    echo 'Run "gist pin <tag1> <tag2>..." to pin tags'
+  else
+    echo Pinned tags: ${pinned_tags[@]/#/#}
+  fi
+}
+
+# pin/unpin tags
+_pin_tags() {
+  # if no arguments, print gists with pinned tags
+  if [[ -z $* && -n $pin ]]; then
+    hint=false _tag_gist $pin
+  else
+    local new_pinned=( $(echo $pin $* | tr ' ' '\n' | sort | uniq -u | xargs) )
+    for tag in ${new_pinned[@]}; do
+      if [[ $tag =~ [p]*[0-9]+ ]]; then 
+        echo Invalid tag: $tag
+        return 1
+      fi
+    done || exit 1
+
+    pin="${new_pinned[@]}"
+    _show_tags
+    sed -i'' -e "/^pin=/ d" "$CONFIG" && echo pin=\'"${new_pinned[@]}"\' >> "$CONFIG"
+  fi
+}
+
+_apply_config "$@" || exit 1
+getConfiguredClient || exit 1
```

Comment:
Can we run checkInternet after this line? It seems all calls will require being able to connect to github.

Reply:
Just part of sub-commands(or sub-commands with proper arguments, like `gist tag <INDEX>`)  need internet, I added `checkInternet` to them.

check 21ce9c7

Reply:
All http request go to `github.com`, so we can just check return code on every http request.
see 77868bb